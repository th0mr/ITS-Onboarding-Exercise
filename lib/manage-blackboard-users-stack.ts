import * as cdk from "aws-cdk-lib";
import type { Construct } from "constructs";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { Tags } from "aws-cdk-lib";
import dotenv from "dotenv";

dotenv.config();

export class ManageBlackboardUsersStack extends cdk.Stack {
    constructor(scope: Construct, id: string, properties?: cdk.StackProps) {
        super(scope, id, properties);

        const lambda = new NodejsFunction(this, "snapshotLambda", {
            functionName: "generateSnapshotFromWebhook",
            entry: "functions/snapshot/generate-snapshot-from-webhook.ts",
            handler: "generateSnapshotFromWebhook",
            runtime: Runtime.NODEJS_20_X,
            memorySize: 128,
            environment: {
                USER_API_ENDPOINT: process.env.USER_API_ENDPOINT ?? "",
            },
        });

        // Add tags to entire stack and all resources
        // Required Tags:
        Tags.of(this).add("york/policy_version", "2");
        Tags.of(this).add("york/name", "ManageBlackboardUserStack");
        Tags.of(this).add("york/group", "ESG");
        Tags.of(this).add("york/project", "ITS-Onboarding-Exercise");
        Tags.of(this).add("york/status", "dev");
        Tags.of(this).add("york/pushed_by", "manual");
        Tags.of(this).add("york/defined_in", "cdk");
        // Optional Tags:
        Tags.of(this).add("york/team", "Teaching and Learning");
        Tags.of(this).add("york/user", "tr901");
    }
}
