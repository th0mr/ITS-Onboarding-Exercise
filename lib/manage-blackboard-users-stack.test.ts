import { describe, it, expect, beforeAll } from "vitest";
import { Template } from "aws-cdk-lib/assertions";
import { App } from "aws-cdk-lib";
import { ManageBlackboardUsersStack } from "./manage-blackboard-users-stack.js";

describe("ManageBlackboardUsersStack", () => {
    beforeAll(() => {
        // Sets up "expect" with a snapshot serializer to replace buckets / assets with [ASSET BUCKET]
        // and [ASSET ZIP] to prevent breaking a snapshot test everytime the source code for the lambda is changed

        // Expected patterns for buckets and asset strings in the snapshots
        const bucketMatch = /cdk-[\da-z]{9}-assets-.*/;
        const assetMatch = /[\da-f]{64}\.zip/;

        expect.addSnapshotSerializer({
            test: (value) =>
                typeof value === "string" &&
                (bucketMatch.exec(value) !== null ||
                    assetMatch.exec(value) !== null),
            print(value) {
                // Substitute both the bucket part and the asset zip part
                let sval = `${value as string}`;
                sval = sval.replace(bucketMatch, "[ASSET BUCKET]");
                sval = sval.replace(assetMatch, "[ASSET ZIP]");
                return `"${sval}"`;
            },
        });
    });

    it("matches the snapshot", () => {
        const stack = new ManageBlackboardUsersStack(
            new App(),
            "ManageBlackboardUsersStack",
        );

        const template = Template.fromStack(stack);

        expect(template).toMatchSnapshot();
    });
});
