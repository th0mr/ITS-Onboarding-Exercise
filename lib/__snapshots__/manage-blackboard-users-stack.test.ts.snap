// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManageBlackboardUsersStack > matches the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "snapshotLambda0B214E59": {
      "DependsOn": [
        "snapshotLambdaServiceRoleDBC47958",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "[ASSET BUCKET]",
          },
          "S3Key": "[ASSET ZIP]",
        },
        "Environment": {
          "Variables": {
            "USER_API_ENDPOINT": "https://webhook.site/e8402416-3786-4d0d-b860-0fa6848b154c",
          },
        },
        "FunctionName": "generateSnapshotFromWebhook",
        "Handler": "index.generateSnapshotFromWebhook",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "snapshotLambdaServiceRoleDBC47958",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "york/defined_in",
            "Value": "cdk",
          },
          {
            "Key": "york/group",
            "Value": "ESG",
          },
          {
            "Key": "york/name",
            "Value": "ManageBlackboardUserStack",
          },
          {
            "Key": "york/policy_version",
            "Value": "2",
          },
          {
            "Key": "york/project",
            "Value": "ITS-Onboarding-Exercise",
          },
          {
            "Key": "york/pushed_by",
            "Value": "manual",
          },
          {
            "Key": "york/status",
            "Value": "dev",
          },
          {
            "Key": "york/team",
            "Value": "Teaching and Learning",
          },
          {
            "Key": "york/user",
            "Value": "tr901",
          },
        ],
      },
      "Type": "AWS::Lambda::Function",
    },
    "snapshotLambdaServiceRoleDBC47958": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "york/defined_in",
            "Value": "cdk",
          },
          {
            "Key": "york/group",
            "Value": "ESG",
          },
          {
            "Key": "york/name",
            "Value": "ManageBlackboardUserStack",
          },
          {
            "Key": "york/policy_version",
            "Value": "2",
          },
          {
            "Key": "york/project",
            "Value": "ITS-Onboarding-Exercise",
          },
          {
            "Key": "york/pushed_by",
            "Value": "manual",
          },
          {
            "Key": "york/status",
            "Value": "dev",
          },
          {
            "Key": "york/team",
            "Value": "Teaching and Learning",
          },
          {
            "Key": "york/user",
            "Value": "tr901",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
